[ENH] Add a general-purpose and emc-related image-handling helper functions

dPys opened the pull request:
The helper functions are included in a new module utils/images.py, which also includes relocated  functions that were previously in included in interfaces/images.py to keep things organized.

pep8speaks commented:
Hello @dPys, thank you for submitting the Pull Request!



* In the file [`dmriprep/utils/images.py`](https://github.com/nipreps/dmriprep/blob/a22eccf551e1a26c48c1d357959b18268c90c09d/dmriprep/utils/images.py):

> [Line 63:100](https://github.com/nipreps/dmriprep/blob/a22eccf551e1a26c48c1d357959b18268c90c09d/dmriprep/utils/images.py#L63): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 99 characters)
> [Line 75:100](https://github.com/nipreps/dmriprep/blob/a22eccf551e1a26c48c1d357959b18268c90c09d/dmriprep/utils/images.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 99 characters)
> [Line 86:100](https://github.com/nipreps/dmriprep/blob/a22eccf551e1a26c48c1d357959b18268c90c09d/dmriprep/utils/images.py#L86): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 99 characters)
> [Line 110:100](https://github.com/nipreps/dmriprep/blob/a22eccf551e1a26c48c1d357959b18268c90c09d/dmriprep/utils/images.py#L110): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 99 characters)
> [Line 184:35](https://github.com/nipreps/dmriprep/blob/a22eccf551e1a26c48c1d357959b18268c90c09d/dmriprep/utils/images.py#L184): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

To test for issues locally, `pip install flake8` and then run `flake8 dmriprep`.

josephmje commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def prune_b0s_from_dwis(in_files, b0_ixs):
+    """Removes B0 volume files from a complete list of dwi volume files."""
```
```suggestion
    """Remove *b0* volume files from a complete list of DWI volume files."""
```

dPys replied:
✔️

josephmje commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def save_3d_to_4d(in_files):
+    """Loads a list of 3D input files and concatenates it to produce a 4D output file."""
```
```suggestion
    """Concatenate a list of 3D volumes into a 4D output."""
```

dPys replied:
✔️

josephmje commented:
Added a bunch of mostly nitpicking comments and made some suggestions to help shorten the line lengths for your docstrings.

josephmje commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+    we already know the rotations and translations matrix,
+    shears and zooms (as outputs from fsl FLIRT/avscale)
+    Let A = the 4x4 rotation and translation matrix
+    R = [          c5*c6,           c5*s6, s5]
```
sphinx is having trouble building this line.

dPys replied:
✔️

dPys replied:
I simplified the docstring anyway so it should build properly now!

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def get_params(A):
+    """This is a copy of spm's spm_imatrix where
```
✔️

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+def match_transforms(dwi_files, transforms, b0_indices):
+    """Arranges the order of a list of affine transforms to correspond with that of each individual dwi volume file,
+    accounting for the indices of B0s. A helper function for EMC."""
+    original_b0_indices = np.array(b0_indices)
```
✔️

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+
+def quick_load_images(image_list, dtype=np.float32):
+    """Iteratively loads 3D dwi volume files from a list of filepaths directly into a 4d array to use for signal
+    prediction. A helper function for EMC."""
```
✔️

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def average_images(images):
+    """Average the voxel-wise signal intensity across a list of 3D image files to produce a 3D mean output image."""
```
✔️

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+    rotations = [Rx, Ry, Rz]
+    translations = [A[0, 3], A[1, 3], A[2, 3]]
+
+    return rotations, translations
```
✔️

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def save_4d_to_3d(in_file):
+    """Loads a 4D input file and splits it in the 4th dimension to produce a list of 3D output files."""
```
✔️

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def quick_load_images(image_list, dtype=np.float32):
+    """Iteratively loads 3D dwi volume files from a list of filepaths directly into a 4d array to use for signal
```
✔️

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,184 @@
+    return out_file
+
+
+def average_images(images):
```
✔️

dPys commented on dmriprep/interfaces/images.py:
```diff
@@ -103,43 +105,26 @@ def _run_interface(self, runtime):
-    return out_file
+    def _run_interface(self, runtime):
+        """
+        Interface for mapping the `match_transforms` function across lists of inputs.
```
✔️

dPys commented on dmriprep/interfaces/images.py:
```diff
@@ -103,43 +105,26 @@ def _run_interface(self, runtime):
+            self.inputs.dwi_files, self.inputs.transforms, self.inputs.b0_indices
+        )
+        return runtime
+    
```
✔️

dPys commented on dmriprep/interfaces/images.py:
```diff
@@ -1,10 +1,12 @@
 """Image tools interfaces."""
 import numpy as np
 import nibabel as nb
+from dmriprep.utils.images import rescale_b0, median, match_transforms
```
✔️

dPys commented:
Thanks @josephmje . I've updated the PR with your review.

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+
+def get_params(A):
+    """Takes an transformation affine matrix A and determines
+    rotations and translations."""
```
Same as above for the docstring:
```suggestion
    """
    Decompose a homogeneous coordinates affine into rotation (3x3) and translations (3x1).

    Parameters
    ----------
    affine : 4x4 :py:class:`numpy.ndarray`
        A transformation affine matrix given in homogeneous coordinates.

    Returns
    -------
    rotations : 3x3 :py:class:`numpy.ndarray`
        A matrix containing the rigid-body rotations corresponding to ``affine``.
    translations : 3x1 :py:class:`numpy.ndarray`
        A vector containing the rigid-body translations component.

    """
```

Please, add some doctest `Examples` to this docstring.

dPys replied:
See above

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+    img_4d = nb.funcs.concat_images([nb.load(img_3d) for img_3d in in_files])
+    out_file = fname_presuffix(in_files[0], suffix="_merged")
+    img_4d.to_filename(out_file)
+    del img_4d
```
This is generally not necessary - unless there is a strong reason, please let the garbage collector do it job.

```suggestion
```

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+    out_file = fname_presuffix(in_files[0], suffix="_merged")
+    img_4d.to_filename(out_file)
+    del img_4d
+    return out_file
```
We should contribute this to Nipype. Why don't we start a module of nibabel interfaces with these functions and also upstream https://github.com/poldracklab/niworkflows/blob/master/niworkflows/interfaces/nibabel.py ?

WDYT @effigies ?

I think the more interfaces with nibabel handling we have screened by Chris, the better.

effigies replied:
I'm not sure why the function exists. It doesn't seem to be being called. And written without unnecessary boilerplate, it's quite short:

```Python
def save_3d_to_4d(in_files):
    out_file = fname_presuffix(in_files[0], suffix="_merged")
    nb.concat_images(in_files).to_filename(out_file)
    return out_file
```

If something this small seems worth upstreaming, sure, why not?

dPys replied:
I agree, this is a nice, general purpose function that could be used in a lot of scenarios. It is basically a pure python equivalent to `fslmerge`?

@effigies -- this function is referenced [here](https://github.com/nipreps/dmriprep/pull/62/files#diff-2ae94ced15c1b8a730ecb0ebdd0f532eR754). We are in the process of incorporating a much larger WIP piecemeal, for which this is a critical helper function.

effigies replied:
If you're using it in a `Function` interface then you need to import `nibabel` in the function body.

dPys replied:
It's included in a general import list, along with numpy and a few others, for all `Function` nodes in the relevant workflow. See: https://github.com/nipreps/dmriprep/pull/62/files#diff-2ae94ced15c1b8a730ecb0ebdd0f532eR692-R701

oesteban replied:
> If something this small seems worth upstreaming, sure, why not?

Yep, I think having nibabel interfaces handy with NiPype would be super-useful for everyone and with the big plus of your eyeballs for the time you keep tabs on nibabel and nipype (hopefully long from my egoistic pov).

dPys replied:
I've kept these in this PR for now, is there a nibabel interface to PR these to in nipype yet?

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+
+def prune_b0s_from_dwis(in_files, b0_ixs):
+    """Remove *b0* volume files from a complete list of DWI volume files."""
+    if in_files[0].endswith("_warped.nii.gz"):
```
This seems to me super-idiosyncratic. Should exist as a private function in the PR where the function is to be used.

dPys replied:
This is also used in signal prediction for EMC, since we don't use the b0's to predict signal-- only dwi volumes. It sounds like the team needs to have a broader conversation RE: our approach to PR-ing all of emc. I have been following the plan outlined in the [six-step outline ](https://github.com/nipreps/dmriprep/pull/62#issuecomment-582020203) posted previously. Is this no longer the approach we want?

arokem replied:
I see now: the file naming convention required is idiosyncratic. That's right. Let's move this out of here and put it into the relevant PR as a private function.

dPys replied:
Cool, I've omitted this for now and plan to add it back later when I PR signal prediction stuff

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+    return out_path
+
+
+def quick_load_images(image_list, dtype=np.float32):
```
Looks like idiosyncratic to some other PR, I would rather add it with that other work.

dPys replied:
This is used in signal prediction for EMC -- loading all images corresponding to the coordinates except for the one left out with LOO. I chose to include it here so that I wouldn't forget about it add since it's a fairly general-purpose image handling-utility. But I could hold back PR-ing this function until the emc signal prediction interface gets added?

dPys replied:
This has now been repurposed into get_list_data -- a more general-purpose function that should hopefully be a little less idiosyncratic.

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+
+def save_4d_to_3d(in_file):
+    """Split a 4D dataset along the last dimension into multiple 3D volumes."""
+    files_3d = nb.four_to_three(nb.load(in_file))
```
This will fail with 3D images. See https://github.com/poldracklab/niworkflows/blob/102f19f50c3c1a5dbf41d2554689356b439f3dae/niworkflows/interfaces/images.py#L62-L71 for reference on how to use nibabel here.

Also, I left another comment about nibabel interfaces/functions.

dPys replied:
Will add some error-handling to this.

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+    return output_matrix
+
+
+def match_transforms(dwi_files, transforms, b0_indices):
```
Why is this necessary? Some examples / unit-tests would be really appreciated.

dPys replied:
Again, this is a helper function for the emc workflows -- its purpose is to prepare an ordered list of affine transform files, corresponding to each respective dwi volume in the series and accounting for the indices of the b0's, to use for subsequent vector rotation.

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+    output_matrix = np.zeros(tuple(example_img.shape) + (num_images,), dtype=dtype)
+    for image_num, image_path in enumerate(image_list):
+        output_matrix[..., image_num] = nb.load(image_path).get_fdata(dtype=dtype)
+    return output_matrix
```
I would let nibabel do this:

```suggestion
    return nb.concat_images([nb.load(fname) for fname in image_list]).get_fdata(dtype=dtype)
```

dPys replied:
yep

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+        nb.squeeze_image(img).to_filename(out_file)
+        return out_file
+
+    median_data = np.median(img.get_fdata(dtype="float32"), axis=-1)
```
I would not force the data type here unless we hit memory issues.
```suggestion
    median_data = np.median(img.get_fdata(), axis=-1)
```

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+
+    hdr = img.header.copy()
+    hdr.set_xyzt_units("mm")
+    hdr.set_data_dtype(np.float32)
```
This will overwrite the original datatype, which is something we probably don't want:
```suggestion
```

We might want to add a `dtype=None` argument to the signature so the data type can be forced if not `None`

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+    return out_file
+
+
+def median(in_file, newpath=None):
```
`newpath` is inconsistent with the `out_path` of `average_images`. I like the behavior of `out_path` below better (i.e., allow the user to set a filename directly, not just rewrite the working directory).

dPys replied:
I agree

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+from nipype.utils.filemanip import fname_presuffix
+
+
+def extract_b0(in_file, b0_ixs, newpath=None):
```
@josephmje could you open a new PR fixing these `newpath` to be consistent to the `out_path` of `average_images` below? I think that behavior is better.

josephmje replied:
I agree. Will do

dPys commented:
> Looks good. I would improve docstrings and adding unit tests would be very much appreciated. Also left some food for thoughts regarding the nibabel manipulations - they'd better be within Nipype.

Awesome, I think I have all the feedback I need now to finish polishing the docstring, and include some doctests.

dPys commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,178 @@
+def match_transforms(dwi_files, transforms, b0_indices):
+    """Arranges the order of a list of affine transforms to correspond with that of
+    each individual dwi volume file, accounting for the indices of B0s. A helper
+    function for EMC."""
```
will clean up the PEP here.

effigies commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,177 @@
+    return out_file
+
+
+def decompose_affine(affine):
```
Is there a reason you're not just using [transforms3d](https://pypi.org/project/transforms3d/)? This is a dependency of niworkflows, so you're probably already installing it.

dPys replied:
No particular reason! I wasn't aware that transforms3d exists. Seems like it might indeed be a nice alternative? What do you think @oesteban ?

oesteban replied:
seems like the best way forward!

dPys commented:
@josephmje -- I went ahead on the latest commit and modified the out_path behavior, but now I see that you have been tackling this in #81 . Would you mind looking at the most recent commit here to see if this is what we want? if so, we can figure out whether we want to indeed just incorporate those changes in #81 or perhaps just keep them here since they are relatively minor tweaks that also now include docstring.

josephmje commented:
> @josephmje -- I went ahead on the latest commit and modified the out_path behavior, but now I see that you have been tackling this in #81 . Would you mind looking at the most recent commit here to see if this is what we want? if so, we can figure out whether we want to indeed just incorporate those changes in #81 or perhaps just keep them here since they are relatively minor tweaks that also now include docstring.

Thanks @dPys ! The docstrings look good. But I don't think the else statement will work if `out_path` is not defined. My PR sets the output path in the interface. Awaiting the CircleCI tests to see if things are working. Are you able to commit to my branch? It would be good to add your docstrings.

dPys commented:
> > @josephmje -- I went ahead on the latest commit and modified the out_path behavior, but now I see that you have been tackling this in #81 . Would you mind looking at the most recent commit here to see if this is what we want? if so, we can figure out whether we want to indeed just incorporate those changes in #81 or perhaps just keep them here since they are relatively minor tweaks that also now include docstring.
> 
> Thanks @dPys ! The docstrings look good. But I don't think the else statement will work if `out_path` is not defined. My PR sets the output path in the interface. Awaiting the CircleCI tests to see if things are working. Are you able to commit to my branch? It would be good to add your docstrings.

Yeah the way you have it in #81 looks good!I plan to work on this tomorrow before we meet so can try to commit my changes, which include docstring and doc-tests, tomorrow morning

dPys commented:
@josephmje @oesteban -- I've added all doctests for this and rebased the PR to the changes already merged in #81, which I noticed was already closed. Folks may want to quickly skim, but assuming all tests pass, this should be okay to merge?

arokem commented on dmriprep/utils/images.py:
```diff
@@ -68,3 +135,232 @@ def median(in_file, dtype=None, out_path=None):
     return out_path
+
+
+def average_images(images, out_path=None):
```
Arguably, for consistency, this should be called "average" (to match "median" above). 

Or better: let's make one function that summarizes images on the last dimension and takes a callable as input. If the callable is `np.median`, you get the median on the last dimension. If it's `np.mean`, then you get the mean. And if your feeling adventurous and pass your own callable here, it gives you back whatever that callable computes, provided it has the same API as `np.median` and `np.mean`, etc. 

@josephmje : what do you think about that?

arokem replied:
I guess we can move ahead with this for now and also revisit later.

dPys replied:
I agree with this. @josephmje What do you think?

josephmje replied:
This sounds good to me. Especially since both functions are nearly identical.

arokem replied:
Now implemented in 7767498

arokem commented on dmriprep/utils/images.py:
```diff
@@ -68,3 +135,232 @@ def median(in_file, dtype=None, out_path=None):
+    ----------
+    file_list : str
+        A list of file paths to 3D NIFTI images.
+
```
```suggestion
Returns 
--------
Nibabel image object
```

josephmje commented on dmriprep/interfaces/images.py:
```diff
@@ -1,13 +1,9 @@
 """Image tools interfaces."""
-from pathlib import Path
-
+from dmriprep.utils.images import rescale_b0, median, match_transforms, extract_b0
```
`pathlib` is being used in the extract and rescale interfaces

```suggestion
from pathlib import Path

```

josephmje commented on dmriprep/interfaces/images.py:
```diff
@@ -1,13 +1,9 @@
+    InputMultiObject, OutputMultiObject
 )
 
 from dmriprep.utils.images import extract_b0, median, rescale_b0
```
```suggestion
from dmriprep.utils.images import extract_b0, median, rescale_b0, match_transforms
```

oesteban commented:
This one seemed reasonably easy to rebase and keep around. Most of this will eventually go to a more general purpose module, but I think there's a net benefit in merging the rebased version. Thanks a lot!
